config/config.go

```diff
@@ -0,0 +1,118 @@
+package config
+
+import (
+	"io/ioutil"
+	"path"
+	"gopkg.in/yaml.v2"
+	"github.com/gruntwork-io/boilerplate/util"
+	"fmt"
+	"github.com/gruntwork-io/boilerplate/errors"
+)
+
+const BOILERPLATE_CONFIG_FILE = "boilerplate.yml"
+
+const OPT_TEMPLATE_FOLDER = "template-folder"
+const OPT_OUTPUT_FOLDER = "output-folder"
+const OPT_NON_INTERACTIVE = "non-interactive"
+const OPT_VAR = "var"
+const OPT_VAR_FILE = "var-file"
+
+// The command-line options for the boilerplate app
+type BoilerplateOptions struct {
+	TemplateFolder 	string
+	OutputFolder 	string
+	NonInteractive	bool
+	Vars		map[string]string
+}
+
+// Validate that the options have reasonable values and return an error if they don't
+func (options *BoilerplateOptions) Validate() error {
+	if options.TemplateFolder == "" {
+		return errors.WithStackTrace(TemplateFolderCannotBeEmpty)
+	}
+
+	if !util.PathExists(options.TemplateFolder) {
+		return errors.WithStackTrace(TemplateFolderDoesNotExist(options.TemplateFolder))
+	}
+
+	if options.OutputFolder == "" {
+		return errors.WithStackTrace(OutputFolderCannotBeEmpty)
+	}
+
+	return nil
+}
+
+// The contents of a boilerplate.yml config file
+type BoilerplateConfig struct {
+	Variables []Variable
+}
+
+// A single variable defined in a boilerplate.yml config file
+type Variable struct {
+	Name 	string
+	Prompt 	string
+	Default	string
+}
+
+// Return the default path for a boilerplate.yml config file in the given folder
+func BoilerPlateConfigPath(templateFolder string) string {
+	return path.Join(templateFolder, BOILERPLATE_CONFIG_FILE)
+}
+
+// Load the boilerplate.yml config contents for the folder specified in the given options
+func LoadBoilerPlateConfig(options *BoilerplateOptions) (*BoilerplateConfig, error) {
+	configPath := BoilerPlateConfigPath(options.TemplateFolder)
+
+	if util.PathExists(configPath) {
+		util.Logger.Printf("Loading boilerplate config from %s", configPath)
+		bytes, err := ioutil.ReadFile(configPath)
+		if err != nil {
+			return nil, err
+		}
+
+		return ParseBoilerplateConfig(bytes)
+	} else {
+		util.Logger.Printf("Warning: boilerplate config file not found at %s. No variables will be available while generating.", configPath)
+		return &BoilerplateConfig{}, nil
+	}
+}
+
+// Parse the given configContents as a boilerplate.yml config file
+func ParseBoilerplateConfig(configContents []byte) (*BoilerplateConfig, error) {
+	boilerplateConfig := &BoilerplateConfig{}
+
+	if err := yaml.Unmarshal(configContents, boilerplateConfig); err != nil {
+		return nil, err
+	}
```

Comment:
Depends on your definition of clean. If you like reflection, then Go's approach is fine. JVM languages have access to the same thing with JSON using a library like Jackson: https://github.com/FasterXML/jackson

For more compile-time driven serialization, Play offers a JSON library (https://www.playframework.com/documentation/2.5.x/ScalaJsonCombinators) that can do some magic via macros. Not sure Go will ever have anything quite like that.

I'm fine with the reflection based approach, so both Go and Java seem equivalent here.